README.md
# Casimir_effect

history.txt
-Нашел эпсилоны подложки и среды, состоящие из 2 лоренцов( c_sub = 16, c_med = 1.7, w0_sub = (0.1,10) , w0_med = (0.05,2) , f_sub = (0.7,0.3) , f_med = (0.8 , 0.29)), у которых 4 точки пересечения. Несколько стабильных положений найти не удалось. Пересечения этих эсилонов слишком близки к 0. Из-за этого мой код плохо считает эту систему, и в любом случае интегралы не уравновешиваются при нескольких d. В планах все такие проверить мой метод проверки интеграла произкодной на 0, обосновать его или указать применимсть. Так же еще поигаться с эпсилонами и улучшить программу, увеличить мелкость сетки мб. Будет очень печально, если придется переписывать код для подсчета двойного интеграла а не одинарного. На этом пока все
-Получил 3 положения равновесия среды и подложки состоящих из 2 осцилляторов лоренца. Нужно теперь подобрать эпсилоны с 4 положениями равновесия, используя 3 осциллятора. Вот осцилляторы в данной работе: (c_sub = 1.8 w0_l_sub = (0.5,10), f_l_sub = (0.9,1.1) , c_med =1.7, w0_l_med = (1.5,w0) , f_l_med = (0.9,0.9)
-Удалось получить 2 стаблиьных положения равновесия. Нужно было просто сделать эпсилон плиты ниже, чем среды и подложки
-Нужно теоретически попробовать доказать применимаость моего метода


stp.py

import sympy as sp
class Transfer_matrix:

	def get_list_of_symbols(self,N,name):
		name = self.system_name + name
		return [sp.symbols(name+str(i),complex =1) for i in range(N)] 

	def get_3d_list_of_symbols(self,N):

		return [sp.zeros(2,2,complex =1) for i in range(N)]
	
	def __init__(self,N,system_name):

		self.system_name = system_name
		
		c = 3*10**(8)

		self.N = N

		self.S_t_matrixs_s = self.get_3d_list_of_symbols(2*N+1) 

		self.S_t_matrixs_p = self.get_3d_list_of_symbols(2*N+1)
 
		self.S_wave_vectors = sp.zeros(1,N+2,complex =1)
 
		self.S_k_z = sp.zeros(1,N+2,complex =1 )

		self.S_permit = self.get_list_of_symbols(N+2,'permit')
		
		self.S_permet = self.get_list_of_symbols(N+2,'permet')

		self.S_length = self.get_list_of_symbols(N,'length')

	def get_r_s_bound(self,k_in_z, k_out_z, p_in, p_out):

		return sp.simplify((p_out*k_in_z - p_in*k_out_z)/(p_out*k_in_z + p_in*k_out_z))

	def get_t_s_bound(self,k_in_z, k_out_z, p_in, p_out):

		return sp.simplify((2*p_out*k_in_z)/(p_out*k_in_z + p_in*k_out_z))

	def get_r_p_bound(self,k_in_z, k_out_z, e_in, e_out):

		return sp.simplify((e_out*k_in_z - e_in*k_out_z)/(e_out*k_in_z + e_in*k_out_z))

	def get_t_p_bound(self,k_in_z, k_out_z, e_in, e_out,p_in,p_out):

		return sp.simplify(2*(e_out*k_in_z)/(e_out*k_in_z + e_in*k_out_z)*sp.sqrt(p_out*e_in/(p_in*e_out)))

	def get_matrix_bound(self,t_in,t_out,r_in,r_out):

		elements = [[t_in*t_out - r_in*r_out, r_out],[(-r_in),1]]
		T = sp.Matrix(elements,complex = 1) 
		return T/t_out

	def get_matrix_slab(self,length,k_z):

		elements = [[sp.exp(length*k_z*1j),0],[0,sp.exp(-length*k_z*1j)]]
		T = sp.Matrix(elements,complex =1) 
		T = sp.simplify(T)
		return T

	def single_det(self,T):
		return(T.row(0)[0]*T.row(1)[1] - T.row(1)[0]*T.row(0)[1])
	def get_multy_det(self,T):   #det(A*B) = det(A)*det(B)
		
		N = self.N
		result = 1
		for i in [j for j in range(2*N+1) if j%2 == 0]: # must be j%2 == 0
			result = result*self.single_det(T[i])

		return result

	def get_t_from_matrix(self,T,T_full):
		return self.get_multy_det(T_full)/T.row(1)[1]

	def get_r_from_matrix(self,T):
		return -T.row(1)[0]/T.row(1)[1]

	def get_coeffs(self,S_w,S_k_x):
		
		N = self.N
		
		S_k = S_w/(3*10**(8))
		
		result = sp.zeros(1,4,complex =1) 

		for i in range(N+2):

			self.S_wave_vectors[i] = sp.simplify(sp.sqrt(self.S_permit[i]*self.S_permet[i])*S_k)

		for i in range(N+2):

			self.S_k_z[i] = sp.simplify(sp.sqrt(self.S_wave_vectors[i]**2 - S_k_x**2))

		for i in range(N+1):	

			t_in_s = self.get_t_s_bound(self.S_k_z[i],self.S_k_z[i+1],self.S_permet[i],self.S_permet[i+1])
			r_in_s = self.get_r_s_bound(self.S_k_z[i],self.S_k_z[i+1],self.S_permet[i],self.S_permet[i+1])
			t_out_s = self.get_t_s_bound(self.S_k_z[i+1],self.S_k_z[i],self.S_permet[i+1],self.S_permet[i])
			r_out_s = self.get_r_s_bound(self.S_k_z[i+1],self.S_k_z[i],self.S_permet[i+1],self.S_permet[i])
			
			t_in_p = self.get_t_p_bound(self.S_k_z[i],self.S_k_z[i+1], self.S_permit[i],self.S_permit[i+1],
																		self.S_permet[i],self.S_permet[i+1])

			r_in_p = self.get_r_p_bound(self.S_k_z[i],self.S_k_z[i+1],self.S_permit[i],self.S_permit[i+1])
			
			t_out_p = self.get_t_p_bound(self.S_k_z[i+1],self.S_k_z[i],self.S_permit[i+1],self.S_permit[i],
																		self.S_permet[i+1],self.S_permet[i])

			r_out_p = self.get_r_p_bound(self.S_k_z[i+1],self.S_k_z[i],self.S_permit[i+1],self.S_permit[i])

			self.S_t_matrixs_s[2*i] = self.get_matrix_bound(t_in_s,t_out_s,r_in_s,r_out_s)
			self.S_t_matrixs_p[2*i] = self.get_matrix_bound(t_in_p,t_out_p,r_in_p,r_out_p)

			if(i < N):

				self.S_t_matrixs_s[2*i+1] = self.get_matrix_slab(self.S_length[i],self.S_k_z[i+1])
				self.S_t_matrixs_p[2*i+1] = self.get_matrix_slab(self.S_length[i],self.S_k_z[i+1])
		
		S_t_result_s = self.S_t_matrixs_s[2*N] # the result S-matrix for all structure
		S_t_result_p = self.S_t_matrixs_p[2*N]
		
		for i in range(1,2*N+1):	
			S_t_result_s = S_t_result_s*self.S_t_matrixs_s[2*N-i]
			S_t_result_p = S_t_result_p*self.S_t_matrixs_p[2*N-i]

		result[0] = self.get_t_from_matrix(S_t_result_s,self.S_t_matrixs_s)		
		result[1] = self.get_r_from_matrix(S_t_result_s)
		result[2] = self.get_t_from_matrix(S_t_result_p,self.S_t_matrixs_p)		
		result[3] = self.get_r_from_matrix(S_t_result_p)

		return result


casimir_m_s_p.py
import numpy as np
import matplotlib
import mpmath as mp
import sympy as sp
import plot_graph
from consts import *
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import stp as tm

class Casimir_m_s_p:
	def __init__(self,args):
		self.eps_1,self.eps_2,self.eps_3 = args #epsilons: s,m,p

	def get_f(self,lib,*params):
		e = sp.symbols('e',positive = 1)
		k = sp.symbols('k',positive = 1)
		d = sp.symbols('d',positive = 1)
		eps = [self.eps_1,self.eps_2,self.eps_3,self.eps_2]
		n1 = 0	
		n2 = 1
		name1 = 'sub'
		name2 = 'plate'
		solver1 = tm.Transfer_matrix(n1,name1)
		solver2 = tm.Transfer_matrix(n2,name2)
		Coeffs1 = solver1.get_coeffs(1j*e,k)
		Coeffs2 = solver2.get_coeffs(1j*e,k)
		r_s1 = Coeffs1[1]
		r_p1 = Coeffs1[3]
		r_s2 = Coeffs2[1]
		r_p2 = Coeffs2[3]
		f_s = (1 - r_s1*r_s2*sp.exp( -2*d*sp.sqrt((e/c)**2*self.eps_2(e) + k**2)))
		f_p = (1 - r_p2*r_p1*sp.exp( -2*d*sp.sqrt((e/c)**2*self.eps_2(e) + k**2)))

		f = ((sp.log(f_s)+sp.log(f_p))*k)

		args = []
		for i in range(n1+2):
			args.append((solver1.S_permit[i],eps[i](e)))
			args.append((solver1.S_permet[i],1))
		for i in range(n2+2):
			args.append((solver2.S_permit[i],eps[i+1](e)))
			args.append((solver2.S_permet[i],1))
		args.append((solver2.S_length[0],5*10**(-9)))
		f = sp.lambdify([(e,k,d,*params)],f.subs(args),lib)
		return f

	def get_deviration(self,lib,*params):
		e = sp.symbols('e',positive = 1)
		k = sp.symbols('k',positive = 1)
		d = sp.symbols('d',positive = 1)
		eps = [self.eps_1,self.eps_2,self.eps_3,self.eps_2]
		n1 = 0	
		n2 = 1
		name1 = 'sub'
		name2 = 'plate'
		solver1 = tm.Transfer_matrix(n1,name1)
		solver2 = tm.Transfer_matrix(n2,name2)
		Coeffs1 = solver1.get_coeffs(1j*e,k)
		Coeffs2 = solver2.get_coeffs(1j*e,k)
		r_s1 = Coeffs1[1]
		r_p1 = Coeffs1[3]
		r_s2 = Coeffs2[1]
		r_p2 = Coeffs2[3]
		kp = sp.sqrt(k**2+(e/c)**2*self.eps_2(e))
		f_s = k*kp*r_s1*r_s2*sp.exp(-2*d*kp)/(1-r_s1*r_s2*sp.exp(-2*d*kp))
		f_p = k*kp*r_p1*r_p2*sp.exp(-2*d*kp)/(1-r_p1*r_p2*sp.exp(-2*d*kp))

		f = f_s + f_p

		args = []
		for i in range(n1+2):
			args.append((solver1.S_permit[i],eps[i](e)))
			args.append((solver1.S_permet[i],1))
		for i in range(n2+2):
			args.append((solver2.S_permit[i],eps[i+1](e)))
			args.append((solver2.S_permet[i],1))
		args.append((solver2.S_length[0],5*10**(-9)))
		f = sp.lambdify([(e,k,d,*params)],f.subs(args),lib)
		return f
	


Casimir_algorithm_v1.py
import numpy as np
import matplotlib
import mpmath as mp
import sympy as sp
import plot_graph
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt
from mpl_toolkits.mplot3d import Axes3D
import casimir_m_s_p as msp
from consts import *

def find_zeros(ar_y,ar_x,epsilon,zero_separation,zeros_count = 10,label = ''):
	if ar_y.shape != ar_x.shape:
		print("error in find_zeros")
		return
	result = []
	vals = np.abs(ar_y)
	j = 0
	for i in range(ar_x.size):
		if len(result) == zeros_count:
			return result	
		if vals[i] < epsilon and j==0:
			result.append(ar_x[i])
			j = zero_separation
		elif j:
			j-=1	
	if(result == []):	 
		return [ar_x[-1]]
	return result
def find_max(Vals,zeros_count):
	Result = np.zeros((Vals.shape[0],zeros_count-1))
	for j in range(Vals.shape[0]):
		row = Temp[j]
		result = []
		for i in range(row.size-2):
			if row[i+1]-row[i] > 0 and row[i+2]-row[i+1] < 0:
				result.append(np.maximum(np.abs(row[i]),0))
			if row[i+1]-row[i] < 0 and row[i+2]-row[i+1] > 0:
				result.append(np.maximum(np.abs(row[i]),0))
		Result[j] = np.array(result)
	Result = np.c_[Temp[:,0],Result]
	return Result

def eps_l(w,f,w0,const):
	w = w*h
	return const + (f[0]*w0[0]**2/(w0[0]**2 + w**2 +w)+ 
                       f[1]*w0[1]**2/(w0[1]**2 + w**2 +w)  )
def eps_d(w):
	return np.exp(4)

def eps_sub(w):
    w0_l_sub = [0.1,sp.symbols('w0')]
    f_l_sub = [0.7,0.3]
    return eps_l(w,f_l_sub,w0_l_sub,1.6)
def eps_med(w):
    w0_l_med= [0.05,2]
    f_l_med = [0.8,0.29]
    return eps_l(w,f_l_med,w0_l_med,1.7)

et_freqs =  np.array([1.62/1000,4.32/1000,1.12/10,6.87,15.2,15.6,43.8])
et_strength =  np.array([9.57/10,1.62,1.4/10,1.26/10,4.16/10,2.44/10,7.1/100])

def eps_et(w):
	w = w*h
	eps = 1.5
	for i in range(7):
		eps += (et_strength[i])/(1 + (w/et_freqs[i])**2)
	return eps
	
solver = msp.Casimir_m_s_p([eps_l_sub,eps_l_med,eps_d])

Deviration = solver.get_deviration('numpy',sp.symbols('w0'))

def Dev(w,k,d,w0):
	return Deviration([w/h,k/t_nm_m,d*t_nm_m,w0])

W = np.array([ 10**(-30) + i/10 for i in range(2000)],dtype=np.cdouble)
D = np.array(list(range(1,100)),dtype = np.cdouble)
W0 = np.array([5 + i for i in range(10)])
All_DATA = np.zeros((W0.size,D.size,W.size))
ALL_ZEROS = []
ALL_MAX = []
ALL_DELTA_ZEROS = []
W_for_eps = np.exp(np.array([-8 + i/50 for i in range(600)]))
Integrlals = np.zeros((W0.size,D.size))
W_,D_ = np.meshgrid(W,D,indexing = 'xy')

w0_l_sub = [0.1,8]
plt.loglog(W_for_eps,eps_l_sub(W_for_eps/h))
plt.loglog(W_for_eps,eps_l_medium(W_for_eps/h))
plt.show()


def eps_diff(w):
	return eps_l_sub(w)-eps_l_medium(w) 

def elmentary_transform(A):
	Temp = np.zeros(A.shape)
	Temp[:,0] = A[:,0]
	for i in range(1,A.shape[1]):
		Temp[:,i] = (-1)**(i)*(A[:,i]-A[:,i-1])	
	return Temp

for j in range(W0.size):
	w0_l_sub[0] = W0[j]
	try:
		vals = eps_diff(W_for_eps/h)
		zeros = find_zeros(vals,W_for_eps,0.05,20) + [0]
		print(zeros)
		Zeros = np.array(zeros*D.size).reshape(D.size,-1)
	except Exception as e:
		raise e
	else:
		Temp = All_DATA[j] = np.real(Dev(W_,0.001,D_,W0[j])) # Temp[i][j] = Dev(W[j],D[i])
		Max = find_max(Temp,Zeros.shape[1])
		ALL_MAX.append(Max)
		for i in range(D.size):
			H = Max[i][-1]/4
			data = Temp[i,W > Zeros[i,-2]+10]
			Zeros[i][-1]=np.real(find_zeros(data,W[W > Zeros[i,-2]+10],H,20,1,str(i))[0])
		ALL_ZEROS.append(Zeros)
		Zeros = elmentary_transform(Zeros)
		ALL_DELTA_ZEROS.append(Zeros)
		Integrlal = np.sum(Zeros*Max,axis = 1).reshape(-1,D.size)
		Integrlals[j] = Integrlal


BRIGHT = 10**(20)
Mask = np.zeros(Integrlals.shape)
enter = input('I\'m ready\n')

def masking(M,DATA,cond):
	try:
		for j in range(M.shape[0]):
			row = DATA[j]
			row_ = M[j]
			temp = []
			for i in range(1,row.size-1):
				if eval(cond):
					temp.append(i)
			row_[temp] = np.array([BRIGHT]*len(temp))
	except Exception as e:
		print(e)	

while enter != 'exit':
	if enter == 'i_t':
		enter = input('...')
		while enter != 'up':
			if enter == 'mask':
				enter = input('...')
				Mask[:,:] = 0
				masking(Mask,Integrlals,enter)
			else:	
				try:
					(i,j),(k,e)= eval(enter)
					plt.matshow(np.maximum(Mask[i:j,k:e],np.abs(Integrlals[i:j,k:e])))
					plt.title('W0 = [' + str(np.round_(W0[0],2)) +','+str(np.round_(W0[-1],2)) + ']')
				except Exception as e:
					print(e)
				else:
					plt.colorbar()
					plt.show()
			enter = input('...')		
	elif enter == 'f_g':
		enter = input('...')
		while enter != 'up':
			try:	
				w0,d = eval(enter)
				plt.plot(np.real(W),All_DATA[w0][d])
				print('...zeros: ',ALL_ZEROS[w0][d])
				print('...maxs: ',ALL_MAX[w0][d])
				print('...delta_zeros: ',ALL_DELTA_ZEROS[w0][d])
			except:
				print('...not 2*int')
			else:
				plt.show()
			enter = input('...')
	enter = input()